client/blocks/qr-code-login/index.jsx

```diff
@@ -0,0 +1,266 @@
+	try {
+		json_response = await response.json();
+	} catch {
+		json_response = false;
+	}
+
+	return json_response;
+};
+
+const fetchQRCodeData = async ( setTokenData, tokenData, anonymousUserId, setIsErrorState ) => {
+	if ( isStillValidToken( tokenData ) ) {
+		return;
+	}
+
+	if ( ! anonymousUserId ) {
+		return;
+	}
+	// tokenData is set to null initially.
+	// Lets wait till it is set to false when the local data is the just yet.
+	if ( tokenData === null ) {
+		return;
+	}
+
+	const responseData = await getLoginActionResponse( 'qr-code-token-request-endpoint', {
+		anon_id: anonymousUserId,
+	} );
+
+	if ( responseData === false ) {
+		setIsErrorState( true );
+		return;
+	}
+
+	if ( isStillValidToken( responseData.data ) ) {
+		setTokenData( responseData.data );
+		setStoredItem( LOCALE_STORAGE_KEY, responseData.data );
+		return;
+	}
+	setIsErrorState( true );
+	return;
+};
+
+const fetchAuthState = async (
+	setAuthState,
+	setTokenData,
+	tokenData,
+	anonymousUserId,
+	isErrorState,
+	setIsErrorState,
+	setPullInterval
+) => {
+	if ( ! tokenData ) {
+		return;
+	}
+
+	if ( isErrorState ) {
+		setPullInterval( null );
+		return;
+	}
+
+	if ( ! anonymousUserId ) {
+		return;
+	}
+
+	if ( ! isStillValidToken( tokenData ) ) {
+		fetchQRCodeData( setTokenData, tokenData, anonymousUserId );
+		return;
+	}
+
+	const { token, encrypted } = tokenData;
+
+	const responseData = await getLoginActionResponse( 'qr-code-authentication-endpoint', {
+		anon_id: anonymousUserId,
+		token,
+		data: encrypted,
+	} );
+
+	if ( responseData === false ) {
+		setIsErrorState( true );
+		setPullInterval( null );
+		return;
+	}
+
+	setAuthState( responseData.data );
+};
+
+function TokenQRCode( { tokenData } ) {
+	if ( ! tokenData ) {
+		return <QRCodePlaceholder />;
+	}
+	const { token, encrypted } = tokenData;
+	const imageSettings = {
+		src: qrCenter,
+		height: 64,
+		width: 64,
+		excavate: false,
+	};
+	return (
+		<QRCode
+			value={ `https://apps.wordpress.com/get/?campaign=login-qr-code#qr-code-login?token=${ token }&data=${ encrypted }` }
+			size={ 300 }
+			imageSettings={ imageSettings }
+		/>
+	);
+}
+
+function QRCodePlaceholder() {
+	return (
+		<div className="qr-code-login__placeholder">
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+		</div>
+	);
+}
+
+function QRCodeLogin() {
+	const translate = useTranslate();
+	const [ tokenData, setTokenData ] = useState( null );
+	const [ authState, setAuthState ] = useState( false );
+	const [ pullInterval, setPullInterval ] = useState( AUTH_PULL_INTERVAL );
+	const [ isErrorState, setIsErrorState ] = useState( false );
+
+	const anonymousUserId = getTracksAnonymousUserId();
+
+	// Set the error state if we don't have a anonymousUserId
+	useEffect( () => {
+		if ( ! anonymousUserId ) {
+			setIsErrorState( true );
+		}
+	}, [ anonymousUserId, setIsErrorState ] );
```

Comment:
`setIsErrorState` doesn't need to be a dependency as it's always constant. The setter functions from `useState` are special and not even the exhaustive-deps ESLint rule won't complain.

Reply:
Thanks for pointing that out.